### PR TITLE
Migration feature

### LINKED ISSUE #18: Handle migration of git repository
If a project gets moved from e.g. GitLab to github the url/port (which is used to identify the project) gets changed and gets treated like a new project. This should be handled by a migration process.

1. Before a new project gets created automatically the user should be asked if he wants to migrate from an other project to this one
2. Rename old project directory
3. Commit renamed directory
4. Edit any existing links
5. Commit links
6. Push changes

### PR DESCRIPTION
**Pull request by eiabea**
Resolves #18

### CODE REVIEW

File: helper/project.ts [hunk 1]
```diff
@@ -130,4 +143,38 @@ export class ProjectHelper {
 
     return parseProjectNameFromGitUrl(originUrl);
   }
+
+  public migrate = async (from: IProject, to: IProject): Promise<void> => {
```
Comment (eiabea) [inline helper/project.ts]: Did i miss some edge cases here? @matthiaseigner @lombad 

- Migrating links will be added in the next pushes

**Comment by eiabea** [comment]
Some functions are untested currently, after writing tests for them the feature should be good to go!

File: helper/project.ts [hunk 2]
```diff
@@ -1,9 +1,40 @@
+import fs, { WriteOptions } from "fs-extra";
 import shelljs, { ExecOutputReturnValue } from "shelljs";
 import uuid from "uuid/v1";
-import { IProject, IRecord } from "../interfaces";
+import { IIntegrationLink, IJiraLink, IProject, IProjectMeta, IRecord } from "../interfaces";
 import { FileHelper, GitHelper, LogHelper, parseProjectNameFromGitUrl } from "./index";
+import { QuestionHelper } from "./question";
 
 export class ProjectHelper {
+  public static projectMetaToDomain = (projectMeta: IProjectMeta): string => {
```
Comment (matthiaseigner) [inline helper/project.ts]: Function descriptions would be nice

File: helper/project.ts (hunk 1 above)
Comment (matthiaseigner) [reply helper/project.ts]: is this already checked @lombad ?

File: helper/project.ts [hunk 3]
```diff
@@ -130,4 +143,38 @@ export class ProjectHelper {
...
   }
+
+  public migrate = async (from: IProject, to: IProject): Promise<void> => {
+    LogHelper.debug("Starting migrate procedure");
+
+    LogHelper.debug(`${from.name} -> ${to.name}`);
+
+    // Ensure all records are present in the "from" project
+    const populatedFrom: IProject | undefined = await this.fileHelper.findProjectByName(from.name, from.meta);
+    if (!populatedFrom) {
```
Comment (lombad) [inline helper/project.ts]: Use `isNullOrUndefined()` here

File: helper/project.ts [hunk 4]
```diff
@@ -130,4 +143,38 @@ export class ProjectHelper {
...
+
+    // TODO check if really the same project object?
+    if (fromDomainProjects.length === 1) {
+      LogHelper.debug(`"from" is only member of domain, remove dir`);
+      // we know that it is not empty, force delete it
+      await this.fileHelper.removeDomainDirectory(from.meta, true);
+    } else {
+      LogHelper.debug(`"from" is not the only member of domain, remove project file`);
+      await this.fileHelper.removeProjectFile(from);
+    }
```
Comment (lombad) [inline helper/project.ts]: As I understand that correctly, this would also work and maybe makes things more clear
```
// delete project File
await this.fileHelper.removeProjectFile(from);

// delete domain directory if empty
const fromDomainProjects: IProject[] = await this.fileHelper.findProjectsForDomain(from.meta);
if (fromDomainProjects.length === 1) {
  LogHelper.debug(`"from" is only member of domain, remove dir`);
  // we know that it is not empty, force delete it
  await this.fileHelper.removeDomainDirectory(from.meta, true);
} 
```

File: helper/project.ts (hunk 3 above)
Comment (eiabea) [reply helper/project.ts]:
```
    /** @deprecated since v4.0.0 - use `value === null || value === undefined` instead. */
    function isNullOrUndefined(object: any): object is null | undefined;
```

beside from that, TS won't recognize this as null checking and claims
```
const project: undefined
Object is possibly 'undefined'.
```

File: helper/project.ts (hunk 4 above)
Comment (eiabea) [reply helper/project.ts]: Implemented your suggestion and added some comments in this [commit](https://github.com/codecovevienna/gittt-cli/pull/19/commits/16fc53cbfbfa2fa3d1f65142df1fb8c169231557)

File: helper/project.ts (hunk 3 above)
Comment (lombad) [reply helper/project.ts]: I actually did not know that and will adopt it in my MR

File: helper/project.ts (hunk 1 above)
Comment (eiabea) [reply helper/project.ts]: Added some TSDoc comments [here](https://github.com/codecovevienna/gittt-cli/pull/19/commits/c07fc38d607b3201b318898228f469bad75c68a4)

File: helper/project.ts (hunk 2 above)
Comment (eiabea) [reply helper/project.ts]: Added some TSDoc comments [here](https://github.com/codecovevienna/gittt-cli/pull/19/commits/c07fc38d607b3201b318898228f469bad75c68a4)

Comment (lombad) [reply helper/project.ts]: Nicely Done :kissing_smiling_eyes:
